expression/aggregation/base_func.go

```diff
@@ -258,6 +269,10 @@ func (a *baseFuncDesc) WrapCastForAggArgs(ctx sessionctx.Context) {
 		panic("should never happen in baseFuncDesc.WrapCastForAggArgs")
 	}
 	for i := range a.Args {
+		// Do not cast the second args of these functions.
+		if i == 1 && (a.Name == ast.WindowFuncLead || a.Name == ast.WindowFuncLag || a.Name == ast.WindowFuncNthValue) {
```

Comment:
Why we need this check?

Reply:
When return type of the function is something other than int, a cast function will make `GetUint64FromConstant` fail and return 0 instead, thus the result will be incorrect.

Reply:
What about adding some test to cover it?

Reply:
This check seems a bit tricky...

Reply:
It is already covered in https://github.com/pingcap/tidb/pull/9672/files#diff-37e6e9e27761faf82eceb330586a5bf2R136